pkg/macaroon/serialize.go

```diff
@@ -0,0 +1,205 @@
+	data = append(data, 2)
+
+	// Serilize Identity
+	data = serializePacket(data, packet{
+		fieldType: fieldIdentifier,
+		data:      m.head,
+	})
+	data = append(data, 0)
+
+	// Serialize caveats
+	for _, cav := range m.caveats {
+		data = serializePacket(data, packet{
+			fieldType: fieldIdentifier,
+			data:      cav,
+		})
+		data = append(data, 0)
+	}
+
+	data = append(data, 0)
+
+	// Serialize tail
+	data = serializePacket(data, packet{
+		fieldType: fieldSignature,
+		data:      m.tail,
+	})
+
+	return data
+}
+
+// serializePacket converts packet to binary
+func serializePacket(data []byte, p packet) []byte {
+	data = appendVarint(data, int(p.fieldType))
+	data = appendVarint(data, len(p.data))
+	data = append(data, p.data...)
+
+	return data
+}
+
+func appendVarint(data []byte, x int) []byte {
+	var buf [binary.MaxVarintLen32]byte
+	n := binary.PutUvarint(buf[:], uint64(x))
+
+	return append(data, buf[:n]...)
+}
+
+// ParseMacaroon converts binary to macaroon
+func ParseMacaroon(data []byte) (*Macaroon, error) {
+	// skip version
+	data = data[1:]
+	// Parse Location
+	data, section, err := parseSection(data)
+	if err != nil {
+		return nil, err
+	}
+	if len(section) > 0 && section[0].fieldType == fieldLocation {
+		section = section[1:]
+	}
+	if len(section) != 1 || section[0].fieldType != fieldIdentifier {
+		return nil, errors.New("invalid macaroon header")
+	}
+
+	mac := Macaroon{}
+	mac.head = section[0].data
+	for {
+		rest, section, err := parseSection(data)
+		if err != nil {
+			return nil, err
+		}
+		data = rest
+		if len(section) == 0 {
+			break
+		}
+		if len(section) > 0 && section[0].fieldType == fieldLocation {
+			section = section[1:]
+		}
+		if len(section) == 0 || section[0].fieldType != fieldIdentifier {
+			return nil, errors.New("no Identifier in caveat")
+		}
+		cav := append([]byte(nil), section[0].data...)
+		section = section[1:]
+		if len(section) == 0 {
+			// First party caveat.
+			//if cav.Location != "" {
+			//	return nil, errors.New("location not allowed in first party caveat")
+			//}
+			mac.caveats = append(mac.caveats, cav)
+			continue
+		}
+		if len(section) != 1 {
+			return nil, errors.New("extra fields found in caveat")
+		}
+		if section[0].fieldType != fieldVerificationID {
+			return nil, errors.New("invalid field found in caveat")
+		}
+		//cav.VerificationId = section[0].data
+		mac.caveats = append(mac.caveats, cav)
+	}
+	data, sig, err := parsePacket(data)
+	if err != nil {
+		return nil, err
+	}
+	if sig.fieldType != fieldSignature {
+		return nil, errors.New("unexpected field found instead of signature")
+	}
+	if len(sig.data) != 32 {
+		return nil, errors.New("signature has unexpected length")
+	}
+	mac.tail = make([]byte, 32)
+	copy(mac.tail[:], sig.data)
+	//return data, nil
+	// Parse Identity
+	// Parse caveats
+	// Parse tail
+	return &mac, nil
+}
+
+// parseSection returns data leftover and packet array
+func parseSection(data []byte) ([]byte, []packet, error) {
```

Comment:
tip: Usually these parsers are much easier to write if you stuff the data into a byte array. 

```
type decoder struct {
    data []byte
    result *Macaroon
}
func (dec *decoder) FieldType() (fieldType, error) { ... }
func (dec *decoder) Int64() (fieldType, error) { ... }
```

But current approach is fine as well.